From 5a5a4ec25ab10bccf567de64c0b6b18bbf012c45 Mon Sep 17 00:00:00 2001
From: Xiao Chen <xiao@apache.org>
Date: Mon, 19 Sep 2016 21:44:42 -0700
Subject: [PATCH 1871/2848] HDFS-10875. Optimize du -x to cache intermediate
 result. Contributed by Xiao Chen.

(cherry picked from commit e52d6e7a46ceef74dd8d8a3d49c49420e3271365)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java

Change-Id: Iaf87a28b87d4f7bd2bdacf4c25b4bec07a1b0934
---
 .../hdfs/server/namenode/INodeDirectory.java       |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java
index d476468..9498064 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java
@@ -608,12 +608,14 @@ public ContentSummaryComputationContext computeContentSummary(int snapshotId,
       ContentSummaryComputationContext summary) {
     final DirectoryWithSnapshotFeature sf = getDirectoryWithSnapshotFeature();
     if (sf != null && snapshotId == Snapshot.CURRENT_STATE_ID) {
+      final Content.Counts counts = Content.Counts.newInstance();
       // if the getContentSummary call is against a non-snapshot path, the
       // computation should include all the deleted files/directories
-      sf.computeContentSummary4Snapshot(summary.getCounts());
-      // Also compute ContentSummary for snapshotCounts (So we can extract it
+      sf.computeContentSummary4Snapshot(counts);
+      summary.getCounts().add(counts);
+      // Also add ContentSummary to snapshotCounts (So we can extract it
       // later from the ContentSummary of all).
-      sf.computeContentSummary4Snapshot(summary.getSnapshotCounts());
+      summary.getSnapshotCounts().add(counts);
     }
     final DirectoryWithQuotaFeature q = getDirectoryWithQuotaFeature();
     if (q != null && snapshotId == Snapshot.CURRENT_STATE_ID) {
-- 
1.7.9.5

